### PR TITLE
Fix availability zones and topology configuration

### PR DESCRIPTION
**Pull request by VariableDeclared**
Fixes 

- https://bugs.launchpad.net/charm-cinder-csi/+bug/2046364
- https://bugs.launchpad.net/charm-cinder-csi/+bug/2059126

### CODE REVIEW

File: charmcraft.yaml [hunk 1]
```diff
@@ -12,13 +12,11 @@ bases:
...
-          - amd64
-          - arm64
+        - "amd64"
       - name: ubuntu
         channel: "22.04"
-        architectures:
-          - amd64
-          - arm64
+        architectures: 
+        - "amd64"
```
Comment (addyess) [inline charmcraft.yaml]: I don't think we need to drop arm64.

File: src/storage_manifests.py [hunk 2]
```diff
@@ -105,6 +105,7 @@ def __init__(self, charm, charm_config, kube_control, integrator):
                 ConfigRegistry(self),
                 CreateStorageClass(self, "default"),  # creates csi-cinder-default
                 UpdateSecrets(self),  # update secrets
+                UpdateControllerPlugin,
```
Comment (addyess) [inline src/storage_manifests.py]: it's gonna need a reference to the base manifest class
```suggestion
                UpdateControllerPlugin(self),
```

File: src/storage_manifests.py [hunk 3]
```diff
@@ -142,3 +143,24 @@ def evaluate(self) -> Optional[str]:
...
+
+    def __call__(self, obj):
+        """Update the controller args in Deployments."""
+        if not any(
+            [
+                (obj.kind == "DaemonSet" and obj.metadata.name == "csi-cinder-nodeplugin"),
+                (obj.kind == "Deployment" and obj.metadata.name == "csi-cinder-controllerplugin"),
+            ]
+        ):
+            return
```
Comment (addyess) [inline src/storage_manifests.py]: My understand was this only affected the "csi-cinder-controllerplugin"

```suggestion
        if not (obj.kind == "Deployment" and obj.metadata.name == "csi-cinder-controllerplugin"):
            return
```

File: src/storage_manifests.py [hunk 4]
```diff
@@ -142,3 +143,24 @@ def evaluate(self) -> Optional[str]:
...
+
+        for container in obj.spec.template.spec.containers:
+            if container.name == "csi-provisioner":
+                for i, val in enumerate(container.args):
+                    if "feature-gates" in val.lower():
+                        container.args[i] = f"feature-gates=Topology={self.manifests.config.get('topology')}"
+                log.info(f"Disabling cinder topology awareness")
```
Comment (addyess) [inline src/storage_manifests.py]: I think that argument needs to be lower-cased

```suggestion
                    if "feature-gates" in val.lower():
                        topology = str(self.manifests.config.get("topology")).lower()
                        container.args[i] = f"feature-gates=Topology={topology}"
                        log.info(f"Configuring cinder topology awareness=%s", topology)
```

File: src/storage_manifests.py [hunk 5]
```diff
@@ -63,12 +63,12 @@ def __call__(self) -> Optional[AnyResource]:
                 metadata=dict(name=storage_name),
                 provisioner="cinder.csi.openstack.org",
                 reclaimPolicy="Delete",
-                volumeBindingMode="WaitForFirstConsumer",
+                volumeBindingMode="WaitForFirstConsumer"
             )
         )
-
+
```
Comment (addyess) [inline src/storage_manifests.py]: LINT: likely a whitespace issue here

File: src/storage_manifests.py [hunk 6]
```diff
@@ -63,12 +63,12 @@ def __call__(self) -> Optional[AnyResource]:
...
                 reclaimPolicy="Delete",
-                volumeBindingMode="WaitForFirstConsumer",
+                volumeBindingMode="WaitForFirstConsumer"
             )
         )
-
+    
         if az := self.manifests.config.get("availability-zone"):
-            sc.parameters.availability = az
+            sc.parameters = dict(availability=az)
```
Comment (addyess) [inline src/storage_manifests.py:70]: What's going on with this?

**Review by addyess — CHANGES_REQUESTED** [review]
Looking VERY nice.  thank you so much
Few nits/questions

File: src/storage_manifests.py (hunk 2 above)
Comment (VariableDeclared) [reply src/storage_manifests.py]: thanks!

File: src/storage_manifests.py (hunk 3 above)
Comment (VariableDeclared) [reply src/storage_manifests.py]: thanks!

File: src/storage_manifests.py (hunk 6 above)
Comment (VariableDeclared) [reply src/storage_manifests.py:70]: The issue is the storage class will lack the parameters object (None), so when trying to fetch the attribute, a accessing of a NoneType error is thrown

File: src/storage_manifests.py (hunk 5 above)
Comment (addyess) [reply src/storage_manifests.py]: Can you remove this whitespace?
```suggestion

```

File: src/storage_manifests.py (hunk 6 above)
Comment (addyess) [reply src/storage_manifests.py:70]: lovely!

File: src/storage_manifests.py [hunk 7]
```diff
@@ -142,3 +143,23 @@ def evaluate(self) -> Optional[str]:
...
+class UpdateControllerPlugin(Patch):
+    """Update the controller args in Deployments."""
+
+    def __call__(self, obj):
+        """Update the controller args in Deployments."""
+        if not any(
+            [
+                (obj.kind == "Deployment" and obj.metadata.name == "csi-cinder-controllerplugin"),
+            ]
+        ):
```
Comment (addyess) [inline src/storage_manifests.py]: It feels like this is easier to read

```suggestion
        if not (obj.kind == "Deployment" and obj.metadata.name == "csi-cinder-controllerplugin"):
```

**Review by addyess — CHANGES_REQUESTED** [review]
Really looking great.  Would you rebase as well since i fixed the linting issue in main

File: src/storage_manifests.py (hunk 7 above)
Comment (VariableDeclared) [reply src/storage_manifests.py]: thanks!

File: src/storage_manifests.py (hunk 5 above)
Comment (VariableDeclared) [reply src/storage_manifests.py]: opps missed that one!

**Comment by VariableDeclared** [comment]
thanks again @addyess !

File: src/storage_manifests.py [hunk 8]
```diff
@@ -142,3 +137,23 @@ def evaluate(self) -> Optional[str]:
             if not self.config.get(prop):
                 return f"Storage manifests waiting for definition of {prop}"
         return None
+
+class UpdateControllerPlugin(Patch):
```
Comment (addyess) [inline src/storage_manifests.py:147]: Human Linter:  Needs 2 lines

```suggestion


class UpdateControllerPlugin(Patch):
```

File: src/storage_manifests.py [hunk 9]
```diff
@@ -142,3 +137,23 @@ def evaluate(self) -> Optional[str]:
...
+        ):
+            return
+
+        for container in obj.spec.template.spec.containers:
+            if container.name == "csi-provisioner":
+                for i, val in enumerate(container.args):
+                    if "feature-gates" in val.lower():
+                        topology = str(self.manifests.config.get("topology")).lower()
+                        container.args[i] = f"feature-gates=Topology={topology}"
+                        log.info(f"Configuring cinder topology awareness=%s", topology)
```
Comment (addyess) [inline src/storage_manifests.py]: Human Linter: f-string is unnecessary and file must end in a newline

```suggestion
                        log.info("Configuring cinder topology awareness=%s", topology)

```

**Review by addyess — CHANGES_REQUESTED** [review]
Hopefully my last nits i swear

**Comment by VariableDeclared** [comment]
thank you again @addyess for your comprehensive review! I really should fix my workpace

File: src/storage_manifests.py (hunk 8 above)
Comment (VariableDeclared) [reply src/storage_manifests.py:147]: thanks!

File: src/storage_manifests.py [hunk 10]
```diff
@@ -77,12 +76,7 @@ class UpdateSecrets(Patch):
 
     def __call__(self, obj):
         """Update the secret volume spec in daemonsets and deployments."""
-        if not any(
-            [
-                (obj.kind == "DaemonSet" and obj.metadata.name == "csi-cinder-nodeplugin"),
-                (obj.kind == "Deployment" and obj.metadata.name == "csi-cinder-controllerplugin"),
-            ]
-        ):
+        if not (obj.kind == "Deployment" and obj.metadata.name == "csi-cinder-controllerplugin"):
```
Comment (addyess) [inline src/storage_manifests.py]: Oops... i didn't do you well if i suggested this line change.  This shouldn't be changed.  It was totally incorrect before
```suggestion
        if not any(
            [
                (obj.kind == "DaemonSet" and obj.metadata.name == "csi-cinder-nodeplugin"),
                (obj.kind == "Deployment" and obj.metadata.name == "csi-cinder-controllerplugin"),
            ]
        ):
```

File: src/storage_manifests.py [hunk 11]
```diff
@@ -142,3 +137,24 @@ def evaluate(self) -> Optional[str]:
...
+class UpdateControllerPlugin(Patch):
+    """Update the controller args in Deployments."""
+
+    def __call__(self, obj):
+        """Update the controller args in Deployments."""
+        if not any(
+            [
+                (obj.kind == "Deployment" and obj.metadata.name == "csi-cinder-controllerplugin"),
+            ]
+        ):
```
Comment (addyess) [inline src/storage_manifests.py]: instead this one only appiles to the controller plugins
```suggestion
        if not (obj.kind == "Deployment" and obj.metadata.name == "csi-cinder-controllerplugin"):
```

**Review by addyess — CHANGES_REQUESTED** [review]
I may have sent you down a bum path unintentionally.  And for that i'm sorry @VariableDeclared

File: src/storage_manifests.py (hunk 10 above)
Comment (VariableDeclared) [reply src/storage_manifests.py]: gosh, my mistake

File: src/storage_manifests.py (hunk 11 above)
Comment (VariableDeclared) [reply src/storage_manifests.py]: yup!

**Comment by VariableDeclared** [comment]
thank you again! made those changes

**Comment by addyess** [comment]
@VariableDeclared 

i'm spotting errors in the unit tests:


```
  File "/home/runner/work/cinder-csi-operator/cinder-csi-operator/tests/unit/test_charm.py", line 162, in test_waits_for_kube_control
    assert storage_messages == {
AssertionError: assert {'Setting secret for Deployment/csi-cinder-controllerplugin', 'Encode secret data for storage.', 'Setting secret for DaemonSet/csi-cinder-nodeplugin', 'Configuring cinder topology awareness=true', 'Creating storage class csi-cinder-default'} == {'Creating storage class csi-cinder-default', 'Setting secret for Deployment/csi-cinder-controllerplugin', 'Setting secret for DaemonSet/csi-cinder-nodeplugin', 'Encode secret data for storage.'}
  
  Extra items in the left set:
  'Configuring cinder topology awareness=true'
  
  Full diff:
    {
  +     'Configuring cinder topology awareness=true',
        'Creating storage class csi-cinder-default',
        'Encode secret data for storage.',
        'Setting secret for DaemonSet/csi-cinder-nodeplugin',
        'Setting secret for Deployment/csi-cinder-controllerplugin',
    }
```

Looks like we need to add following line 162 of `/tests/unit/test_charm.py`

``` 'Configuring cinder topology awareness=true'```

**Review by addyess — APPROVED** [review]
Hooray!  🍰 
LGTM

**Pull request merged**